### DISCUSSION TITLE
[@types/google-spreadsheet] GoogleSpreadsheet constructor bad type definition

### DISCUSSION THREAD

**Discussion by diego-alfadev** [DefinitelyTyped/DefinitelyTyped#66005] [Issues with a @types package] [answered]
Im working on a typescript code with the google-spreadsheet and @types/google-spreadsheet libraries and i found a mismatch between the library type definition and source library docs.

As explained [here](https://theoephraim.github.io/node-google-spreadsheet/#/guides/authentication?id=service-account), `GoogleSpreadsheet` class can be instanciated directly wit a `JWT` instance (from `google-auth-library`):

```javascript

import { JWT } from 'google-auth-library'

import creds from './config/myapp-1dd646d7c2af.json'; // the file saved above

const SCOPES = [
  'https://www.googleapis.com/auth/spreadsheets',
  'https://www.googleapis.com/auth/drive.file',
];

const jwt = new JWT({
  email: creds.client_email,
  key: creds.private_key,
  scopes: SCOPES,
});
const doc = new GoogleSpreadsheet('<YOUR-DOC-ID>', jwt);

```

But in mi code, this wont be accepted by typescript type-checking: 

```typescript

public static getDoc = async (docId: DocIdType) => {

        let credentials: ServiceAccountCredentials

        if (docId in GoogleSheetsHelper.CREDENTIALS) {
            credentials = GoogleSheetsHelper.CREDENTIALS[docId]
        }

        const jwt = new JWT({
            email: credentials.client_email,
            key: credentials.private_key,
            scopes: SCOPES,
        });

        // This will throw a compilaition error "Expected 0-1 arguments, but got 2"
        const doc = new GoogleSpreadsheet(docId, jwt);

```

Checking the type definition source, this message is aligned with the definition: 

```typescript
export class GoogleSpreadsheet implements SpreadsheetBasicProperties {
    /**
     * @description
     * create a new Spreadsheet doc
     *
     * @param sheetId document ID from the URL of the Spreadsheet
     */
    constructor(sheetId?: string);
...
```

Im aware theres other method i can use to authenticate, such as using `doc.useServiceAccountAuth({client_email, private_key})` later on, but since that approach is not the first use case scenario in the docs i'll recomend updating the `GoogleSpreadsheet` accordingly, just for others to not waste time searching for that approach or using `@ts-ignore`. Adding an optional jwt parameter will be enough.

In the other hand, maybe the constructor approach is being deprecated in favor of `useServiceAccountAuth` but i didn't see anything in that direction in API documentation.

Thanks in advance.

[a comment by an automated account is not shown]

**Comment by EndBug** [comment, marked as answer]
Hey Diego, thanks for bringing this up

We're actually deprecating `@types/google-spreadsheet` in favor of the built-in types provided by `google-spreadsheet@^4.0.0`

Related:
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65874
- https://github.com/theoephraim/node-google-spreadsheet/issues/622

**Reply by diego-alfadev** [reply]
Thanks Fede, i'm glad to hear this.
My issue was that npm installed 3.3.0 version by default, reason why i needed the @types dependency.
Upgrading to 4.0.0 here. Thanks for your response.
